AP_Notify & AP_AHRS: Final changes for Solo to operate on Master

Pedals2Paddles opened the pull request:
This PR contains all the outstanding notify related issues that inhibit Solo from operating cleanly on ArduCopter.  All compiles and tests ok on copter.  **With these commits merged into master, the solo will no longer need a custom version or dedicated build target.**

* **TONES:** Removed the solo specific tones.  There wasn't much different. And if it will be on ArduPilot, it should sound like ArduPilot.  Solo will use the standard px4 tones.  Added the power off tone to the px4 tone file.  The only tone that no longer exists is the "gps unplugged" tone, which really only served to annoy the user while servicing the vehicle. 

* **NOTIFY DEVICE SELECTION:** Jaime did the hard work on this one.  He created `NTF_BRD_TYPE` and
reworked notify device selection to take place on init rather than on compile like before.  Notify device selection is now completely dynamic so the user can select what applies to them and not bother with what doesn't apply. This should save some memory here and there too. Thank you Jaime!!  I then modified it a bit further, creating `NTF_OLED_THEME`.  This will allow the user to enable/disable the OreoLED driver. And it also allows you to select between aircraft and rover lighting themes. This allows the Solo to use the OreoLEDs, and doesn't waste the memory on vehicles not equipped with Oreo LEDs.  

* **GPS FUSION FLAG:** I created a new  notify flag for GPS Fusion in AP_Notify.  This flag is set true by the EKF filter state in `AP_AHRS_NavEKF.cpp` when the EKF is happy with the GPS, actively using it for position information. It is set false when the EKF is rejecting the GPS.  This flag is used for visual operator notification of GPS status.

* **OREO LED REWORK:** Reworked the process flow.  Created the aviation and rover themes. Created visual feedback for prearm checks, GPS, EKF, and gyro init, Modified visual feedback for radio failsafe, and low battery. Also brought in Hugh Eaves' awesome work to allow overriding the Oreo LEDs via mavlink.

If it isn't too late to get this into the 3.5 release, that would be really awesome.  This would allow a Solo to operate on the production release of master without any special versions or builds. It would be a significant milestone for the project!

Pedals2Paddles commented:
@jmachuca77 thank you again for the rework on notify device selection. We test flew it as you created it and it is worked great.  I made some modifications for the Oreo LEDs and to account for removing the solo tone files. 

@hugheaves thank you again for the OreoLED extended settings and mavlink control!

tridge commented on libraries/AP_Notify/OreoLED_PX4.h:
```diff
@@ -74,10 +110,73 @@ class OreoLED_PX4 : public NotifyDevice
+            clear_state();
+        }
+
+        inline void clear_state() {
```
please remove the inline calls and instead move these functions to the cpp file

Pedals2Paddles replied:
Got it.
@hugheaves, we need to rework this a bit.

hugheaves replied:
Will do. Also, l'll fix the byte ordering on those uint16 values too. Never realized (until I looked at the constants you'd added) that I had the byte order backwards. (duh!)

Pedals2Paddles commented:
@tridge & @jmachuca77, I reworked the rework as discussed in the dev call.  Removed the `NTF_BRD_TYPE` parameter.  Used preprocessor directives to set everything based on HAL board types and subtypes.  I kept Jaime's new structure for setting the devices since it works nicely even with preprocessor directives.  Retained the `NTF_OLED_THEME` parameter to switch the Oreo LED driver on and off, and will only allow it on Pixhawk 2 board type.  Compiled and tested.

@hugheaves is reworking the OreoLED for the inline function thing as discussed on the dev call. Coming soon on that.

Pedals2Paddles commented:
Yay.  All checks finally passed on `AP_Notify.cpp` changes.

Pedals2Paddles commented:
@tridge @rmackay9 @proficnc 
All changes requested in the dev call have been completed.  Rebased, squashed, and all checks passed!  Can this be merged and backported to 3.5?

rmackay9 commented on libraries/AP_Notify/AP_Notify.cpp:
```diff
@@ -78,93 +85,145 @@ AP_Notify::AP_Notify()
+        _devices[1] = new ToshibaLED_I2C();
+        _devices[2] = new ToneAlarm_PX4();
+        _devices[3] = new ExternalLED();
+        _devices[4] = new Display();
```
It may be fine but I guess adding the "Display" is a small functional change for the VRBRAIN boards?

rmackay9 commented on libraries/AP_Notify/AP_Notify.h:
```diff
@@ -138,10 +140,18 @@ class AP_Notify
     AP_Int8 _rgb_led_override;
     AP_Int8 _buzzer_enable;
     AP_Int8 _display_type;
+    AP_Int8 _oled_theme;
```
To be consistent with the parameter name change suggested above, maybe we should rename this member to _oreo_theme as well? I'm just thing we can be consistent in using "oreo" instead of using both "oreo" and "oled".

rmackay9 commented on libraries/AP_Notify/AP_Notify.h:
```diff
@@ -138,10 +140,18 @@ class AP_Notify
     uint32_t _send_text_updated_millis; // last time text changed
     char _flight_mode_str[5];
+    
+    NotifyDevice* boardled;
```
No need for these pointers any more I think (boardled, toshibaled, etc).

rmackay9 commented on libraries/AP_Notify/OreoLED_PX4.h:
```diff
@@ -89,6 +142,12 @@ class OreoLED_PX4 : public NotifyDevice
     oreo_state _state_sent[OREOLED_NUM_LEDS];       // last state sent to led
     uint8_t _pattern_override;                      // holds last processed pattern override, 0 if we are not overriding a pattern
+    uint8_t _oled_theme;
+    uint8_t rear_color_r{255};                           // the rear LED red value
```
our normal convention to initialise the variables is to use "=" unless it's an array.  So we would use, "rear_color_r = 255" instead of "rear_color_r{255}".
Also perhaps we should be consistent about using an unscore before the variable name?  We don't actually mind whether we use it or not, but let's be consistent within a class or file.  So perhaps, "rear_color_r" should be renamed to "_rear_color_r".

Pedals2Paddles commented:
@rmackay9 I am not seeing what in copter EKF Check would drive a flag specific to GPS?  It appears to make it do that, would require the same modification to the AHRS library to begin with. The copter EKF_Check would be a (seemingly) unnecessary middle man. Having in the AHRS library lets it be specific to GPS, and it supports all vehicles. 

I will fix up the rest of the items you pointed out.  I agree on all of those, it will be much cleaner and more consistent.

Pedals2Paddles commented:
@rmackay9 I fixed up all the code as discussed. It is definitely much cleaner (and shorter) now. Squashed the commits and updated the comments so those are nice and clean now too.  All checks have passed!

rmackay9 commented:
I've had another look and found some of the formatting changes were still missing so I've had an attempt at making them here: https://github.com/rmackay9/rmackay9-ardupilot/commits/matt-solo2

I couldn't quite complete the review, hopefully tomorrow.

rmackay9 commented:
By the way, I still think the gps update should be moved into the vehicle code.  Although it's an edge case, imagine if both EKF2 and EKF3 were running.  both would update the flag.

Pedals2Paddles commented:
In the AHRS library as currently written, only one EKF can ever be active. AHRS_EKF_USE is determining which one is in use, and that is the only one that can set the flag. There is nothing in the vehicle code to detect what I am detecting. To do that, best I can tell, would require exactly what I'm using in the AHRS library to drive the flag directly anyway. And it would then not be available on other vehicles. Maybe I'm missing something here. 

I didn't know you wanted all code revised to remove all camel case. I only edited the ones you called out in the PR comments. Is there something that details stuff like that? Had I known camel case was frowned upon in ArduPilot, I would have avoided it. I wish I knew that before.

Pedals2Paddles commented:
I took in the last of those formatting changes and rebased. All checks have passed.

rmackay9 commented:
I've looked at it a bit more, a few final changes merged into my branch here.  i'll let travis test it and then merge to master. https://github.com/rmackay9/rmackay9-ardupilot/commits/matt-solo3

rmackay9 commented:
Merged!
Congratulations @Pedals2Paddles, @jmachuca77 and I think one more person (Hughes?) contributed to this PR!
In any case, congrats!

Pedals2Paddles commented:
Thank you everyone!
